### ISSUE TITLE
Adapt simple-recurrence-network to regression

### ISSUE THREAD

**Issue by MLBenjii** [Element-Research/rnn#284]
Hello,

I am new to Torch / NN, and so I am trying to create a simple regression model based on the simple-recurrence-network example as a learning exercise.

I thought this would be as simple as:
- changing the lookup table to a linear layer
- set nIndex to 1 (overload nIndex as input size)
- remove LogSoftMax at output
- use MSECriterion in place of the ClassNLLCriterion

However, when I run this model, I get an error from the first Linear module referenced by the ParallelTable:

Linear.lua:57: size mismatch, [7 x 1], [8]

My hidden layer is size 7, and batch size is 8. This should be doing a matrix-vector mult of [7x1] matrix by [1] vector, but instead the batch is being passed through (by the sequencer?).

I assume I am missing something very basic here.... any help appreciated!

Setup of model is as follows:
local rm = nn.Sequential() 
   :add(nn.ParallelTable()
      :add(nn.Linear(nIndex, hiddenSize)) 
      :add(nn.Linear(hiddenSize, hiddenSize))) 
   :add(nn.CAddTable()) -- merge
   :add(nn.Sigmoid()) -- transfer

local rnn = nn.Sequential()
   :add(nn.Recurrence(rm, hiddenSize, 0)) 
   :add(nn.Linear(hiddenSize, nIndex))

rnn = nn.Sequencer(rnn)

**Comment by rracinskij** [comment]
Hi @MLBenjii ,
can't replicate your error. What is your input? Why do you use ParallelTable?

**Comment by MLBenjii** [comment]
Hi Richard,

Isn't the parallel table needed with the recurrence module in order to feedback previous output into the model? I am tweaking the recurrence classification model, which does use a parallel table:

https://github.com/Element-Research/rnn/blob/master/examples/simple-recurrence-network.lua

For inputs, I'll upload the full src when I'm back at my keyboard, but essentially same as the sample except real valued types instead of integers, but still trying to predict the next number in the sequence (eg: predict x+1 given x)

Thx!

Get Outlook for Android<https://aka.ms/ghei36>



On Fri, Jul 1, 2016 at 12:28 AM -0400, "Ricard Racinskij" <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @MLBenjii<https://github.com/MLBenjii> ,
can't replicate your error. What is your input? Why do you use ParallelTable?

-

**Comment by rracinskij** [comment]
@MLBenjii , I may be wrong, but it looks like that the problem is caused by different dimensionality of nn.Linear and nn.LookupTable. Your input is probably a table of tensors of size 8 and following code shows the difference:

````
local nIndex = 1
local hiddenSize = 7

local input1 = torch.Tensor(8):fill(1)
local input2 = torch.Tensor(8,1):fill(1)
local net1 = nn.Sequential():add(nn.LookupTable(nIndex, hiddenSize))
local net2 = nn.Sequential():add(nn.Linear(nIndex, hiddenSize))

print(net1:forward(input1)) --[torch.DoubleTensor of size 8x7]
print(net1:forward(input2)) --[torch.DoubleTensor of size 8x1x7]
print(net2:forward(input2)) --[torch.DoubleTensor of size 8x7] as nn.Linear takes 1 input
print(net2:forward(input1)) --size mismatch, [7 x 1], [8] as you are trying to use a Tensor(8) as nn.Linear(1,7) input
````
Unfortunately, I have no experience with rnn.Recurrence, so I have no immediate hint on how to change your network setup, but hope this helps.

Other question is why don't you use the https://github.com/Element-Research/rnn/blob/master/examples/recurrent-time-series.lua example?

**Comment by nicholas-leonard** [comment]
Sorry for delay. The input must be 8 x 1 (this is a difference between Linear and LookupTable) and the 3rd argument of is also now 1 instead of 0 as it represents the number of non-batch dimensions"

```lua
nIndex = 1
hiddenSize = 7
batchSize = 8

local rm = nn.Sequential()
:add(nn.ParallelTable()
:add(nn.Linear(nIndex, hiddenSize))
:add(nn.Linear(hiddenSize, hiddenSize)))
:add(nn.CAddTable()) -- merge
:add(nn.Sigmoid()) -- transfer

local rnn = nn.Sequential()
:add(nn.Recurrence(rm, hiddenSize, 1))
:add(nn.Linear(hiddenSize, nIndex))

rnn = nn.Sequencer(rnn)


input = {torch.randn(batchSize, nIndex), torch.randn(batchSize, nIndex)}
rnn:forward(input)
```
